src/python/WMCore/Storage/RucioFileCatalog.py

```diff
@@ -0,0 +1,236 @@
+#!/usr/bin/env python
+"""
+_RucioFileCatalog_
+
+Object to contain LFN to PFN mappings from a Rucio File Catalog
+and provide functionality to match LFNs against them
+
+Usage:
+
+given a storage description file, storage.json, execute readRFC. This will return
+a RucioFileCatalog instance that can be used to match LFNs to PFNs.
+"""
+
+from builtins import next, str, range
+from future.utils import viewitems
+
+from future import standard_library
+standard_library.install_aliases()
+
+import os
+import re
+
+import json
+
+from urllib.parse import urlsplit
+from xml.dom.minidom import Document
+
+from WMCore.Algorithms.ParseXMLFile import xmlFileToNode
+
+class RucioFileCatalog(dict):
+    """
+    _RucioFileCatalog_
+
+    Object that can map LFNs to PFNs based on contents of a Rucio
+    File Catalog
+    """
+
+    def __init__(self):
+        dict.__init__(self)
+        self['lfn-to-pfn'] = []
+        self['pfn-to-lfn'] = []
+        self.preferredProtocol = None  # attribute for preferred protocol
+
+    def addMapping(self, protocol, match, result,
+                   chain=None, mapping_type='lfn-to-pfn'):
+        """
+        _addMapping_
+
+        Add an lfn to pfn mapping to this instance
+
+        """
+        entry = {}
+        entry.setdefault("protocol", protocol)
+        entry.setdefault("path-match-expr", re.compile(match))
+        entry.setdefault("path-match", match)
+        entry.setdefault("result", result)
+        entry.setdefault("chain", chain)
+        self[mapping_type].append(entry)
+
+    def _doMatch(self, protocol, path, style, caller):
+        """
+        Generalised way of building up the mappings.         
+        :param protocol: the name of a protocol, for example XRootD
+        :path: a LFN path, for example /store/abc/xyz.root
+        :style: type of conversion. lfn-to-pfn is to convert LFN to PFN and pfn-to-pfn is for PFN to LFN
+        :caller is the method from there this method was called. It's used for resolving chained rules
+        """
+        for mapping in self[style]:
+            if mapping['protocol'] != protocol:
+                continue
+            if mapping['path-match-expr'].match(path) or mapping["chain"] != None:
+                if mapping["chain"] != None:
+                    oldpath = path
+                    path = caller(mapping["chain"], path)
```

Comment:
At some point, it will stop because all the `chain` are traced back to the root. This condition `if mapping["chain"] != None:` is used to break the circular loop (the root mapping with not contain `chain` attributes).
These code are copied from TrivialFileCatalog (master branch) without any modification, except the doc string:
https://github.com/nhduongvn/WMCore/blob/87ea437f508308b5e1d544a234a80e2a9911d5c1/src/python/WMCore/Storage/TrivialFileCatalog.py#L75
Here is an example of `chain`
https://gitlab.cern.ch/SITECONF/T1_DE_KIT/-/blob/master/PhEDEx/storage_disk.xml?ref_type=heads#L24
You can see the root mapping is `pnfs` which does not have `chain` attributes
https://gitlab.cern.ch/SITECONF/T1_DE_KIT/-/blob/master/PhEDEx/storage_disk.xml?ref_type=heads#L15
and the loop stops here